src/buildstream_plugins/sources/cargo.py

```diff
@@ -294,6 +324,284 @@ def _get_mirror_file(self, sha=None):
+                depth=1,
+            )
+
+    ########################################################
+    #        Helper APIs for the Cargo Source to use       #
+    ########################################################
+
+    def ref_node(self):
+        return {"kind": "git", "name": self.name, "version": self.version, "repo": self.repo, "commit": self.commit}
+
+    # stage()
+    #
+    # A delegate method to do the work for a single git repo
+    # in Source.stage().
+    #
+    # Args:
+    #    (directory): The vendor subdirectory to stage to
+    #
+    def stage(self, directory):
+        self.cargo.status(f"Checking out {self.commit}")
+
+        crate_target_dir = os.path.join(directory, f"{self.name}-{self.version}")
+        tmp_dir = os.path.join(directory, f"{self.name}-{self.version}-tmp")
+
+        try:
+            os.mkdir(tmp_dir)
+        except FileExistsError:
+            raise SourceError(
+                f"This project requests crate {self.name} {self.version} from multiple sources, "
+                "which is incompatible with vendoring since cargo does not support it."
+            )
+
+        with Repo(self._get_mirror_dir(), bare=True) as mirror:
+            with Repo.init(tmp_dir) as dest:
+                dest.object_store.add_object(mirror[self.commit.encode()])
+                dest.refs[b"HEAD"] = self.commit.encode()
+                dest.update_shallow([self.commit.encode()], [])
+
+            with Repo(tmp_dir, object_store=mirror.object_store) as dest:
+                dest.reset_index()
+
+        # Workspace handling
+        #
+        # When new workspace features are added it is worth checking if
+        # <https://github.com/flatpak/flatpak-builder-tools/blob/HEAD/cargo/flatpak-cargo-generator.py>
+        # has implemented them already. This implementation is inspired by the mentioned source.
+
+        with open(os.path.join(tmp_dir, "Cargo.toml"), "rb") as f:
+            root_toml = tomllib.load(f)
+
+        crates = {}
+
+        if "workspace" in root_toml:
+            # Find wanted crate inside workspace
+            for member in root_toml["workspace"].get("members", []):
+                for crate_toml_path in glob.glob(os.path.join(tmp_dir, member, "Cargo.toml")):
+                    crate_path = os.path.normpath(os.path.dirname(crate_toml_path))
+
+                    with open(crate_toml_path, "rb") as f:
+                        crate_toml = tomllib.load(f)
+                        crates[crate_toml["package"]["name"]] = {
+                            "config": crate_toml,
+                            "path": crate_path,
+                        }
+
+            crate = crates[self.name]
+            # Apply information inherited from workspace Cargo.toml
+            config_inherit_workspace(crate["config"], root_toml["workspace"])
+
+            with open(os.path.join(crates[self.name]["path"], "Cargo.toml"), "bw") as f:
```

Comment:
I suppose this should be "w", not "bw" based on https://gitlab.gnome.org/GNOME/gnome-build-meta/-/merge_requests/2825#note_2092004

Reply:
That was fallout from switching from tomli-w to tomlkit. It's fixed now.

>  I suggest using text mode everywhere for consistency.

tomllib that's used for loading does only support binary mode

Reply:
This is not true based on brief glance of source code. It seems to support both. Maybe it's not true on some older version?

Reply:
> Read a TOML file. The first argument should be a readable and binary file object.

https://docs.python.org/3/library/tomllib.html